src/sentry/api/endpoints/project_rules.py

```diff
@@ -44,66 +48,219 @@ def pre_save_rule(instance, sender, *args, **kwargs):
     clean_rule_data(instance.data.get("actions", []))
 
 
+@dataclass
+class MatcherResult:
+    has_key: bool = False
+    key_matches: bool = False
+
+
+class DuplicateRuleEvaluator:
+    ACTIONS_KEY = "actions"
+    ENVIRONMENT_KEY = "environment"
+    SPECIAL_FIELDS = [ACTIONS_KEY, ENVIRONMENT_KEY]
+
+    EXCLUDED_FIELDS = ["name", "user_id"]
+
+    def __init__(
+        self,
+        project_id: int,
+        rule_data: dict[Any, Any] | None = None,
+        rule_id: int | None = None,
+        rule: Rule | None = None,
+    ) -> None:
+        """
+        rule.data will supersede rule_data if passed in
+        """
+        self._project_id: int = project_id
+        self._rule_data: dict[Any, Any] = rule.data if rule else rule_data
+        self._rule_id: int | None = rule_id
+        self._rule: Rule | None = rule
+
+        self._keys_to_check: set[str] = self._get_keys_to_check()
+
+        self._matcher_funcs_by_key: dict[str, Callable[[Rule, str], MatcherResult]] = {
+            self.ENVIRONMENT_KEY: self._environment_matcher,
+            self.ACTIONS_KEY: self._actions_matcher,
+        }
+
+    def _get_keys_to_check(self) -> set[str]:
+        """
+        Returns a set of keys that should be checked against all existing rules.
+        Some keys are ignored as they are not part of the logic.
+        Some keys are required to check, and are added on top.
+        """
+        keys_to_check: set[str] = {
+            key for key in list(self._rule_data.keys()) if key not in self.EXCLUDED_FIELDS
+        }
+        keys_to_check.update(self.SPECIAL_FIELDS)
+
+        return keys_to_check
+
+    def _get_func_to_call(self, key_to_check: str) -> Callable:
+        return self._matcher_funcs_by_key.get(key_to_check, self._default_matcher)
+
+    def _default_matcher(self, existing_rule: Rule, key_to_check: str) -> MatcherResult:
+        """
+        Default function that checks if the key exists in both rules for comparison, and compares the values.
+        """
+        match_results = MatcherResult()
+
+        existing_rule_key_data = existing_rule.data.get(key_to_check)
+        current_rule_key_data = self._rule_data.get(key_to_check)
+        if existing_rule_key_data and current_rule_key_data:
+            match_results.has_key = True
+
+        if match_results.has_key:
+            match_results.key_matches = existing_rule_key_data == current_rule_key_data
```

Comment:
You can move this to line 111 and get rid of has_key

Reply:
We can't get rid of `has_key` that easily, see conversation at: https://github.com/getsentry/sentry/pull/64922/files#r1488478610